pallets/xyk-liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,931 @@
+		///
+		/// Emits:
+		/// * `RewardClaimed` if claim happen
+		/// * `SharesWithdrawn` event when successful
+		#[pallet::weight(<T as Config>::WeightInfo::withdraw_shares())]
+		pub fn withdraw_shares(
+			origin: OriginFor<T>,
+			deposit_id: DepositId,
+			yield_farm_id: YieldFarmId,
+			asset_pair: AssetPair,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let nft_owner = T::NFTHandler::owner(&T::NftCollectionId::get(), &deposit_id)
+				.ok_or(Error::<T>::CantFindDepositOwner)?;
+
+			ensure!(nft_owner == who, Error::<T>::NotDepositOwner);
+
+			let global_farm_id = T::LiquidityMiningHandler::get_global_farm_id(deposit_id, yield_farm_id)
+				.ok_or(Error::<T>::DepositDataNotFound)?;
+
+			let amm_pool_id = T::AMM::get_pair_id(asset_pair);
+
+			let unclaimable_rewards: Balance = if T::LiquidityMiningHandler::is_yield_farm_claimable(
+				global_farm_id,
+				yield_farm_id,
+				amm_pool_id.clone(),
+			) {
+				//This should not fail on double claim, we need unclaimable_rewards
+				let fail_on_double_claim = false;
+				let (global_farm_id, reward_currency, claimed, unclaimable) = T::LiquidityMiningHandler::claim_rewards(
+					who.clone(),
+					deposit_id,
+					yield_farm_id,
+					fail_on_double_claim,
+				)?;
+
+				if !claimed.is_zero() {
+					Self::deposit_event(Event::RewardClaimed {
+						global_farm_id,
+						yield_farm_id,
+						who: who.clone(),
+						claimed,
+						reward_currency,
+						deposit_id,
+					});
+				}
+
+				unclaimable
+			} else {
+				0
+			};
+
+			let (global_farm_id, withdrawn_amount, is_destroyed) =
+				T::LiquidityMiningHandler::withdraw_lp_shares(deposit_id, yield_farm_id, unclaimable_rewards)?;
+
+			let lp_token = Self::get_lp_token(&amm_pool_id)?;
+			if !withdrawn_amount.is_zero() {
+				Self::deposit_event(Event::SharesWithdrawn {
+					global_farm_id,
+					yield_farm_id,
+					who: who.clone(),
+					lp_token,
+					amount: withdrawn_amount,
+					deposit_id,
+				});
+			}
+
+			if is_destroyed {
+				Self::unlock_lp_tokens(lp_token, &who, withdrawn_amount)?;
+				T::NFTHandler::burn(&T::NftCollectionId::get(), &deposit_id, Some(&nft_owner))?;
+
+				Self::deposit_event(Event::DepositDestroyed { who, deposit_id });
+			}
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Account ID of the pot holding locked LP shares. This account is also owner of NFT class
+	/// for all the NFTs minted by this pallet.
+	fn account_id() -> Option<T::AccountId> {
+		<T as pallet::Config>::PalletId::get().try_into_account()
+	}
+
+	fn get_lp_token(amm_pool_id: &T::AccountId) -> Result<AssetId, Error<T>> {
+		let assets = T::AMM::get_pool_assets(amm_pool_id).ok_or(Error::<T>::CantGetAmmAssets)?;
+		let asset_pair = AssetPair::new(assets[0], assets[1]);
+
+		//NOTE: this check is important AMM:get_share_token() return `0` if amm doesn't exist
+		if !T::AMM::exists(asset_pair) {
+			return Err(Error::<T>::AmmPoolDoesNotExist);
+		}
```

Comment:
I would expect
```suggestion
		ensure!(T::AMM::exists(asset_pair), Error::<T>::AmmPoolDoesNotExist);
```
to work here.
Maybe you would need to adjust the Error in the result above to `DispatchError`, not sure.